src/main/kotlin/biokotlin/kmer/KmerIO.kt

```diff
@@ -0,0 +1,225 @@
+class KmerIO(filename: String, isCompressed: Boolean = true): Iterator<Pair<Kmer, Int>> {
+    private val reader: BufferedReader
+
+    private var sequenceLength: Long? = null
+    private var ambiguousKmers: Long? = null
+    private var setType: String? = null
+    private var stepSize: Int? = null
+    private var keepMinOnly: Boolean? = null
+    private var bothStrands: Boolean? = null
+    private var kmerSize: Int? = null
+    private var currentLine: String? = ""
+
+    init {
+        assert(filename.isNotEmpty())
+
+        reader = if (isCompressed) {
+            BufferedReader(InputStreamReader(LZ4FrameInputStream(FileInputStream(File(filename)))))
+        } else {
+            File(filename).bufferedReader()
+        }
+
+        // first line should be a header
+        val firstLine = reader.readLine()
+        if(!firstLine.startsWith("#")) { throw FileNotFoundException("File does not contain a header line")}
+
+        // map header to set values
+        val split = firstLine.substring(1).split(",")
+        split.forEach{
+            val pair = it.split(":").map{str -> str.trim()}
+            when(pair[0]) {
+                "sequenceLength" -> sequenceLength = pair[1].toLong()
+                "ambiguousKmers" -> ambiguousKmers = pair[1].toLong()
+                "setType" -> setType = pair[1]
+                "stepSize" -> stepSize = pair[1].toInt()
+                "keepMinOnly" -> keepMinOnly = pair[1] == "true"
+                "bothStrands" -> bothStrands = pair[1] == "true"
+                "kmerSize" -> kmerSize = pair[1].toInt()
+            }
+        }
+
+        // check for required parameters, print warning if missing
+        if (sequenceLength == null) {println("Warning: header does not specify sequence length.")}
+        if (ambiguousKmers == null) {println("Warning: header does not specify ambiguous kmers.")}
+        if (setType == null) {println("Warning: header does not specify set type.")}
+        if (stepSize == null) {println("Warning: header does not specify step size.")}
+        if (keepMinOnly == null) {println("Warning: header does not specify keepMinOnly.")}
+        if (bothStrands == null) {println("Warning: header does not specify bothStrands.")}
+        if (kmerSize == null) {println("Warning: header does not specify kmer size.")}
+
+        if(!listOf("KmerSet", "KmerBigSet", "KmerMultiSet").contains(setType)) { println("Warning: header does not specify a recognized set type.") }
+
+        loadNextLine()
+    }
+
+    /**
+     * Returns an AbstractKmerSet of all kmers in the file.
+     * The header must specify a valid set type, or an exception will be thrown.
+     */
+    fun readAll(): AbstractKmerSet {
+        return when (setType) {
+            "KmerBigSet" -> {readBigSet()}
+            "KmerSet" -> {readKmerSet()}
+            "KmerMultiSet" -> {readKmerMultiSet()}
+            else -> {throw FileNotFoundException("Provided file does not describe a recognized set type")}
```

Comment:
Same comment as above - FileNotFoundException seems misleading.  A different error with your provided comment, perhaps augmented with details on possible value for set type?